[#11][#12][#13] admin-api 모듈 추가

dvlprkoji opened the pull request:
[#11] 회원 가입 기능 추가
[#12] 쿠폰 등록 기능 추가
[#13] 이벤트 등록 기능 추가
[#15] 쿠폰 수량 업데이트 기능 추가

dvlprkoji committed:
feat: 회원 가입 기능 추가

Closes #11

dvlprkoji committed:
feat: 쿠폰 등록 기능 추가

Closes #12

dvlprkoji committed:
feat: 이벤트 등록 기능 추가

Closes #13

dvlprkoji committed:
feat: 쿠폰 수량 업데이트 기능 추가

Closes #15

dvlprkoji committed:
refactor: 패키지 구조 변경

- request, response 패키지 aladiin-core 모듈로 이동

dvlprkoji committed:
feat: 이벤트 참여 모듈 추가

- 이벤트 참여 상태 조회 기능 추가
- 이벤트 참여 기능 추가

dvlprkoji committed:
test: 이벤트 참여 상태 조회 기능 테스트 추가

f-lab-lyan commented on admin-api/src/main/java/aladiin/adminapi/service/EventService.java:
```diff
@@ -1,20 +1,38 @@
+        saveCouponStock(event);
+    }
+
+    private void saveCouponStock(Event event) {
```
이 코드는 couponStock에 대한 작업만 하는 코드로 보이는데, EventService에 있어야 하는 이유가 있을까요? 혹은 couponStockRepository는 계속 EventService에서만 사용되는 것인가요?

dvlprkoji replied:
이 코드가 EventService 밖의 Controller 단에 있어야 한다는 말씀이실까요?
couponStockRepository 는

1. 이벤트 등록시 쿠폰 재고 채우기
2. 쿠폰 발급 과정에서 쿠폰 재고 감소시키기

정도 역할을 할 것 같네요

f-lab-lyan commented on aladiin-core/src/main/java/aladiin/core/config/RedisConfiguration.java:
```diff
@@ -0,0 +1,37 @@
+public class RedisConfiguration {
+
+
+    @Value("${spring.redis.host}")
```
@ConfigurationProperties?

dvlprkoji replied:
말씀해주신 방법이 더 좋아보이네요 수정하겠습니다

f-lab-lyan commented on aladiin-core/src/main/java/aladiin/core/domain/entity/CouponStock.java:
```diff
@@ -0,0 +1,23 @@
+    private int quantity;
+
+    public static CouponStock of(Event event, LocalDate eventDate) {
+        String key = "event:" + event.getId()
```
이 키가 그대로 DB에 저장되는 것 같은데요. 이 키를 이대로 저장하면 어떤 이득을 취할 수 있는것인가요?

dvlprkoji replied:
고민해보지 않은 부분입니다.. 어떤 방식이 있을까요?

f-lab-lyan commented on aladiin-core/src/main/java/aladiin/core/domain/entity/dao/CouponStockRepository.java:
```diff
@@ -0,0 +1,21 @@
+
+
+    public void save(CouponStock couponStock) {
+        RAtomicLong atomicLong = redissonClient.getAtomicLong(couponStock.getKey());
+        atomicLong.set(couponStock.getQuantity());
```
이건 하나의 트랜젝션으로 묶어서 처리하는게 더 좋을 것 같은데요?

dvlprkoji replied:
네 트랜잭션으로 묶어서 처리해야겠네요

f-lab-lyan commented on coupon-api/src/main/java/aladiin/couponapi/CouponApiApplication.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication(scanBasePackages = {"aladiin.couponapi", "aladiin.core"})
```
그냥 같은 패키지로 만들면, 이 설정이 필요없지 않을까요? :-)

dvlprkoji replied:
별도로 어플리케이션을 띄울 생각으로 모듈을 분리했습니다!

f-lab-lyan commented on coupon-api/src/main/java/aladiin/couponapi/controller/EventController.java:
```diff
@@ -0,0 +1,39 @@
+
+    @GetMapping("/{eventId}/status")
+    private ResponseEntity<CommonResponse> getJoinStatus(@PathVariable Long eventId,
+                                                         @RequestParam(name = "eventdate") String eventDate,
```
위에도 비슷한 코드가 있었는데요. 이건 같은 이벤트에 대해서 여러날의 이벤트가 있을 수 있다는 것을 가정한 것인가요?

dvlprkoji replied:
네 같은 이벤트에 대해 여러 날의 이벤트가 있을 수 있을 것으로 가정했습니다

f-lab-lyan commented on coupon-api/src/main/java/aladiin/couponapi/exception/InvalidDiscountRatioException.java:
```diff
@@ -0,0 +1,9 @@
+package aladiin.couponapi.exception;
+
+public class InvalidDiscountRatioException extends InvalidDiscountException {
```
이건 InvalidDiscountException을 상속받은 특별한 이유가 있나요?

dvlprkoji replied:
InvalidDiscountException 을 catch 해서 별도의 처리를 할 수 있을 것 같아서 작성해 보았습니다.

f-lab-lyan commented on coupon-api/src/main/java/aladiin/couponapi/kafka/EventJoinProducer.java:
```diff
@@ -0,0 +1,27 @@
+        log.info("produce [{}] into broker [{}]", eventJoinDTO, broker);
+        kafkaTemplate.send(topic, eventJoinDTO);
+    }
+}
```
Please add a new line at the end of the file.

f-lab-lyan commented on coupon-api/src/test/java/aladiin/couponapi/controller/EventControllerTest.java:
```diff
@@ -0,0 +1,85 @@
+import static org.assertj.core.api.Assertions.*;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@RequiredArgsConstructor
```
이 annotation이 여기에 필요한 것인가요?

dvlprkoji replied:
네 필요 없을 것 같네요 테스트 파일을 계속 수정하다가 지우지 않고 남긴 것 같네요

f-lab-lyan commented on coupon-api/src/test/java/aladiin/couponapi/controller/EventControllerTest.java:
```diff
@@ -0,0 +1,85 @@
+                .baseUrl(baseUrl)
+                .build();
+
+        LocalDateTime today = LocalDateTime.of(LocalDate.now(), LocalTime.MIDNIGHT);
```
이 날짜가 꼭 오늘일 필요가 있나요? 테스트를 진행할때?

dvlprkoji replied:
이벤트 시작일과 이벤트 종료일 사이의 랜덤 값으로 정하면 될 것 같네요
굳이 오늘로 정할 필요는 없을 것 같습니다

dvlprkoji committed:
refactor: CouponStock 관련 메소드 EventService 에서 couponStockService 로 이동

dvlprkoji committed:
refactor: CommonResponse 클래스 패키지 이동

dvlprkoji committed:
refactor: @Value 대신 @ConfigurationProperties 사용

dvlprkoji committed:
refactor: CouponStock 관련 메소드 EventService 에서 couponStockService 로 이동

dvlprkoji committed:
feat: testcontainers 라이브러리 적용

dvlprkoji committed:
test: admin-api 모듈 EventControllerTest 추가

- TestContainer 적용

dvlprkoji committed:
refactor: JPA 테이블 키 생성전략 변경 (AUTO > IDENTITY)

dvlprkoji committed:
refactor: EventJoinMember 변경

- Redis 자료구조 변경 (Strings > Sets)
- key 값 변경 ("event:{eventId}:date:{eventDate}:member:{memberId}" > "{eventId}:{eventDate}")

dvlprkoji committed:
test: coupon-api 모듈 EventControllerTest 추가

dvlprkoji committed:
refactor: @Value 대신 @ConfigurationProperties 사용

dvlprkoji committed:
fix: Profile 명 수정 (dev > !test)

dvlprkoji committed:
refactor: @Value 대신 @ConfigurationProperties 사용

dvlprkoji committed:
fix: 테스트 초기정보 @BeforeEach 에서 주입하도록 수정

dvlprkoji committed:
refactor: Kafka message key 로 memberId 를 설정

dvlprkoji committed:
fix: Kafka Producer 설정 변경

- 브로커 1개 > 3개 변경
